### PR TITLE
Add string concatenation node and update ir tree visitor

### PR DESCRIPTION
**Pull request by jdconrad**
Two changes:
1. The first is the addition of a specialized string concatenation node as part of the ir tree. This allows us to roll up all of the binary nodes containing string concatenations into one node as a set of arguments which will happen in the next PR. This will help in removing optimization complexity from the semantic phases.
2. This updates the ir tree visitor to have a base class with visitChildren for each node making new ir phases easier to write as only the nodes that are relevant need to be visited. This change is mostly mechanical with the addition of a visitChildren to each ir node.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ArgumentsNode.java [hunk 1]
```diff
@@ -36,6 +36,6 @@ public void addArgumentNode(ExpressionNode argumentNode) {
         return argumentNodes;
     }
 
-    /* ---- end tree structure */
+    /* ---- end tree structure, visitor ---- */
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ArgumentsNode.java:39]: Copy/Paste error

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java [hunk 2]
```diff
@@ -44,10 +44,17 @@ public Location getLocation() {
...
      */
-    public <Input, Output> Output visit(IRTreeVisitor<Input, Output> irTreeVisitor, Input input) {
+    public <Scope> void visit(IRTreeVisitor<Scope> irTreeVisitor, Scope scope) {
         throw new UnsupportedOperationException("cannot visit ir node type [" + getClass().getCanonicalName() + "]");
     }
 
+    /**
+     * Visits all child ir tree nodes for this ir tree node.
+     */
+    public <Scope> void visitChildren(IRTreeVisitor<Scope> irTreevisitor, Scope scope) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:54]: Can this be declared abstract?

**Review by stu-elastic — APPROVED** [review]
Make the unsupported ops into abstract where appropriate.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ArgumentsNode.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ArgumentsNode.java:39]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java]: Yes, changed. Good call.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Will commit as soon as CI passes.

**Pull request merged**